### PR TITLE
Introduce "mixed" protocol

### PR DESCRIPTION
**Pull request by smatting**
This is a partial implementation of

https://wearezeta.atlassian.net/browse/FS-1574
https://wearezeta.atlassian.net/browse/FS-1576

to unblock client development.

## Checklist

 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)

### CODE REVIEW

File: libs/wire-api/src/Wire/API/Conversation/Protocol.hs [hunk 1]
```diff
@@ -144,3 +159,15 @@ deriving via (Schema Protocol) instance ToJSON Protocol
 protocolDataSchema :: ProtocolTag -> ObjectSchema SwaggerDoc Protocol
 protocolDataSchema ProtocolProteusTag = tag _ProtocolProteus (pure ())
 protocolDataSchema ProtocolMLSTag = tag _ProtocolMLS mlsDataSchema
+protocolDataSchema ProtocolMixedTag = tag _ProtocolMixed mlsDataSchema
+
+newtype ProtocolUpdate = ProtocolUpdate {unProtocolUpdate :: ProtocolTag}
```
Comment (pcapriotti) [inline libs/wire-api/src/Wire/API/Conversation/Protocol.hs:164]: For future-proofing, maybe we should wrap the JSON instance in an object.

File: services/galley/src/Galley/API/Update.hs [hunk 2]
```diff
@@ -679,6 +684,53 @@ checkReusableCode convCode = do
...
+updateLocalConversationProtocol qusr client _mconn lcnv (ProtocolUpdate newProtocol) = do
+  conv <- E.getConversation (tUnqualified lcnv) >>= noteS @'ConvNotFound
+  void $ ensureOtherMember lcnv qusr conv
+  case (protocolTag (convProtocol conv), newProtocol) of
+    (ProtocolProteusTag, ProtocolMixedTag) -> do
+      E.updateToMixedProtocol lcnv MLS_128_DHKEMX25519_AES128GCM_SHA256_Ed25519
+      E.addMLSClients (convToGroupId lcnv) qusr (Set.singleton (client, nullKeyPackageRef))
+    (ProtocolMixedTag, ProtocolMixedTag) ->
+      pure ()
+    (_, _) -> throwS @'ConvInvalidProtocolTransition
```
Comment (pcapriotti) [inline services/galley/src/Galley/API/Update.hs:732]: For consistency, maybe we can accept transitions from protocol `p` to `p`.

**Review by pcapriotti — APPROVED** [review]
Looks good. Just minor comments.

File: libs/wire-api/src/Wire/API/Conversation/Protocol.hs [hunk 3]
```diff
@@ -114,15 +119,25 @@ protocolValidAction (ProtocolMLS _) ConversationLeaveTag = True
 protocolValidAction (ProtocolMLS _) ConversationRemoveMembersTag = False
 protocolValidAction (ProtocolMLS _) ConversationDeleteTag = True
 protocolValidAction (ProtocolMLS _) _ = True
+protocolValidAction (ProtocolMixed _) ConversationJoinTag = error "TODO"
+protocolValidAction (ProtocolMixed _) ConversationLeaveTag = error "TODO"
+protocolValidAction (ProtocolMixed _) ConversationRemoveMembersTag = error "TODO"
+protocolValidAction (ProtocolMixed _) ConversationDeleteTag = error "TODO"
+protocolValidAction (ProtocolMixed _) _ = error "TODO"
```
[the remark this answers is not in the document]
Comment (smatting) [reply libs/wire-api/src/Wire/API/Conversation/Protocol.hs]: Right, it was a leftover TODO

File: libs/wire-api/src/Wire/API/Conversation/Protocol.hs (hunk 1 above)
Comment (smatting) [reply libs/wire-api/src/Wire/API/Conversation/Protocol.hs:160]: That was my intention, even with the newtype wrapper. But I made a mistake in the schema definition. Fixed now

File: services/galley/test/integration/API/MLS.hs [hunk 4]
```diff
@@ -3526,3 +3531,78 @@ testCreatorRemovesUserFromParent = do
               )
               (sort [alice1, charlie1, charlie2])
               (sort $ pscMembers sub2)
+
+testMixedUpgrade :: TestM ()
```
[the remark this answers is not in the document]
Comment (smatting) [reply services/galley/test/integration/API/MLS.hs:3535]: Yes, but it wasn't ready yet and I didn't want to be blocked by it.

File: services/galley/src/Galley/API/Update.hs (hunk 2 above)
Comment (smatting) [reply services/galley/src/Galley/API/Update.hs:732]: Makes sense. I've added it

**Pull request merged**